src/System.Private.CoreLib/src/System/Diagnostics/Eventing/EventPipePayloadDecoder.cs

```diff
@@ -0,0 +1,183 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+using System.Reflection;
+using System.Runtime.InteropServices;
+
+namespace System.Diagnostics.Tracing
+{
+#if FEATURE_PERFTRACING
+    internal static class EventPipePayloadDecoder
+    {
+        /// <summary>
+        /// Given the metadata for an event and an event payload, decode and deserialize the event payload.
+        /// </summary>
+        internal static object[] DecodePayload(ref EventSource.EventMetadata metadata, ReadOnlySpan<Byte> payload)
```

Comment:
I must admit that body of the loop looks roughly like DecodeObject (but the parameters are problematic).   It would be nice to have only one decoder.  Make a DecodeObject(Type, byte*) that can be called both places.     They are really doing the same thing (taking a blob and converting it to objects so that EventListener can be called.   We really should not have two pieces of code.

Can you at least make an issue about it?

Reply:
Yes, they are similar.  I am trying to remember why I chose to separate them, but I admit, I can't remember the details.  I have filed https://github.com/dotnet/coreclr/issues/18752 to track merging them.